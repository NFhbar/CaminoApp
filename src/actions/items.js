import uuid from 'uuid';

// ADD_ITEM
export const addItem = (
  {
    description = '',
    note = '',
    amount = 0,
    createdAt = 0
  } = {}
) => ({
  type: 'ADD_ITEM',
  item: {
    id: uuid(),
    description,
    note,
    amount,
    createdAt
  }
});

// REMOVE_ITEM
export const removeItem = ({ id } = {}) => ({
  type: 'REMOVE_ITEM',
  id
});

// EDIT_ITEM
export const editItem = (id, updates) => ({
  type: 'EDIT_ITEM',
  id,
  updates
});
