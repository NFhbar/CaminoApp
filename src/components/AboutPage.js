import React from 'react';
import { NavLink } from 'react-router-dom';
import { row, column } from 'reactstrap';

const AboutPage = () => (
  <div className="row card-body">
    <div className="col-sm-12">
      <div className="jumbotron">
        <h1 className="display-4">React-Redux Camino Financial App</h1>
        <p className="lead">This App is a take-home assignment built by
          <a href="https://www.linkedin.com/in/nicolas-frega/"> Nicolas Frega </a>
          for <a href="https://www.caminofinancial.com/"> Camino Financial </a>.
         </p>
        <hr className="my-4"/>
        <p>The App is built with React, Redux, Bootstrap, and Chartjs. It is deployed using Heroku. It mimics some
        of the behavior of a store App in which the admin can add items to sell at a
        given price. Added items are randomly generated a sold amount for displaying purposes. Items can be sorted by price or amount sold.
        The items are then rendered dynamically to the page. The charts are statics
        since no real sales data is present.</p>
        <p className="lead">
          <a className="btn btn-primary btn-lg" href="https://github.com/NFhbar/CaminoApp" role="button">View github</a>
        </p>
      </div>
    </div>
  </div>
);

export default AboutPage;
