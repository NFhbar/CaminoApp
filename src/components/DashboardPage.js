import React from 'react';
import { row, column } from 'reactstrap';
import ItemList from './ItemList';
import ItemPlot from './ItemPlot';
import ItemListFilters from './ItemListFilters';

const DashboardPage = () => (
  <div className="row card-body">
    <div className="col-sm-12">
      <ItemListFilters />
      <ul className="list-group">
        <li className="list-group-item disabled">
          <div className="row">
            <div className="col-sm-4">
              Item
            </div>
            <div className="col-sm-4">
              Price
            </div>
            <div className="col-sm-4">
              Items sold
            </div>
          </div>
        </li>
        <ItemList />
      </ul>
      <ItemPlot />
    </div>
  </div>
);

export default DashboardPage;
