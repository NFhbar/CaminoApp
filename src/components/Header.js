import React from 'react';
import { NavLink } from 'react-router-dom';
import { row, column } from 'reactstrap';

const Header = () => (
  <div className="row card-header">
    <div className="col-sm-12">
        <h1 className="card-title" id="header-title">Camino Financial React App</h1>
        <div className="row">
          <div className="col-md-4">
            <NavLink to="/" className="btn btn-primary" activeClassName="is-active" exact={true}>Dashboard</NavLink>
          </div>
          <div className="col-md-4">
            <NavLink to="/create" className="btn btn-primary" activeClassName="is-active">Add Item</NavLink>
          </div>
          <div className="col-md-4">
            <NavLink to="/about" className="btn btn-primary" activeClassName="is-active">About</NavLink>
          </div>
        </div>
    </div>
  </div>
);

export default Header;
