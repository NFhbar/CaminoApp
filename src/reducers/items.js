// Items Reducer

const itemsReducerDefaultState = [];

export default (state = itemsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_ITEM':
      return [
        ...state,
        action.item
      ];
    case 'REMOVE_ITEM':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_ITEM':
      return state.map((item) => {
        if(item.id === action.id){
          return {
            ...item,
            ...action.updates
          };
        } else {
          return item;
        };
      });
    default:
      return state;
  }
};
