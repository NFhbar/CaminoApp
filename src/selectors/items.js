// Get visible items
export default (items, {text, sortBy}) => {
  return items.filter((item) => {
    const textMatch = item.description.toLowerCase().includes(text.toLowerCase());

    return textMatch;
  }).sort((a, b) => {
    if (sortBy === 'amount') {
      return a.amount < b.amount ? 1: -1;
    } else if (sortBy === 'sales') {
      return a.sales < b.sales ? 1: -1;
    }
  });
};
