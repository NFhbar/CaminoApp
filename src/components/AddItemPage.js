import React from 'react';
import { connect } from 'react-redux';
import { row, column } from 'reactstrap';
import ItemForm from './ItemForm'
import { addItem } from '../actions/items';

export class AddItemPage extends React.Component {
  onSubmit = (item) => {
    this.props.addItem(item);
    this.props.history.push('/');
  };
  render() {
    return (
      <div className="row card-body">
        <div className="col-sm-12">
          <h3>Add an item to your store!</h3>
          <ItemForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  addItem: (item) => dispatch(addItem(item))
});

export default connect(undefined, mapDispatchToProps)(AddItemPage);
