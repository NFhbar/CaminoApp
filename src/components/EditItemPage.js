import React from 'react';
import { connect } from 'react-redux';
import { row, column } from 'reactstrap';
import ItemForm from './ItemForm';
import { editItem, startRemoveItem } from '../actions/items';

export class EditItemPage extends React.Component {
  onSubmit = (item) => {
    this.props.editItem(this.props.item.id, item);
    this.props.history.push('/');
  };
  onRemove = () => {
    this.props.startRemoveItem({ id: this.props.item.id });
    this.props.history.push('/');
  };
  render() {
    return (
      <div className="row" id="edit-item-page">
        <div className="col-sm-12">
          <ItemForm
            item={this.props.item}
            onSubmit={this.onSubmit}
          />
          <button type="submit" id="btn-remove" className="btn btn-md btn-danger" onClick={this.onRemove}>Remove Item</button>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state, props) => ({
  item: state.items.find((item) => item.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  editItem: (id, item) => dispatch(editItem(id, item)),
  startRemoveItem: (data) => dispatch(startRemoveItem(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditItemPage);
