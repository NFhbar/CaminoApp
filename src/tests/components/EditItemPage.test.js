import React from 'react';
import { shallow } from 'enzyme';
import items from '../fixtures/items';
import { EditItemPage } from '../../components/EditItemPage';

let startEditItem, startRemoveItem, history, wrapper;

beforeEach(() => {
  startEditItem = jest.fn();
  startRemoveItem = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditItemPage
      startEditItem={startEditItem}
      startRemoveItem={startRemoveItem}
      history={history}
      item={items[2]}
    />
  );
});

test('should render EditItemPage', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle startEditItem', () => {
  wrapper.find('ItemForm').prop('onSubmit')(items[2]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startEditItem).toHaveBeenLastCalledWith(items[2].id, items[2]);
});

test('should handle startRemoveItem', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startRemoveItem).toHaveBeenLastCalledWith({
    id: items[2].id
  });
});
