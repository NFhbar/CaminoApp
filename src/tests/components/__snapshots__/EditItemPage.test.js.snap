// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditItemPage 1`] = `
<div
  className="row"
  id="edit-item-page"
>
  <div
    className="col-sm-12"
  >
    <ItemForm
      item={
        Object {
          "amount": 1500,
          "description": "Coat",
          "id": "3",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="btn btn-md btn-danger"
      id="btn-remove"
      onClick={[Function]}
      type="submit"
    >
      Remove Item
    </button>
  </div>
</div>
`;
